### PR TITLE
Replace `flake8` with `ruff`

### PR DESCRIPTION
**Pull request by EarlMilktea**
Related to #8481 .

This PR replaces `flake8` integrated in `.pre-commit-config.yaml` with `ruff`.

### Changes

- create `.ruff.toml` mostly compatible with the current linting configuration
- fix/silence linter warnings that only `ruff` can detect
- fix invalid `noqa` directives

### Note

- we don't have to update the contribution guide as the details of `pre-commit` CI are intentionally hidden

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -21,11 +21,14 @@ repos:
...
+    rev: v0.6.9
+    hooks:
+      - id: ruff
+
   - repo: https://github.com/PyCQA/flake8
     rev: 7.1.1
     hooks:
-      - id: flake8
-        # use config from setup.cfg
       - id: flake8
```
Comment (kmaehashi) [inline .pre-commit-config.yaml:32]: I think it's now OK to remove flake8 against Cython code as well. We've introduced cython-lint over a year ago (#7508) and it's working perfectly.

File: .ruff.toml [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+  "cupy/array_api",
+  "docs/source",
+  "tests/cupy_tests/array_api_tests",
+  "third_party",
+]
+
+[lint]
+select = ["E", "F", "W"]
+ignore = [
+  "E501", # line-too-long (let formatters handle this)
```
Comment (kmaehashi) [inline .ruff.toml:14]: Is it ok to ignore this?

**Review by kmaehashi — COMMENTED** [review]
LGTM except for comments!

File: .pre-commit-config.yaml (hunk 1 above)
Comment (EarlMilktea) [reply .pre-commit-config.yaml]: Resolved in f62318a866963b3501e7c3f67b0917645df547c7 .

File: .ruff.toml (hunk 2 above)
Comment (EarlMilktea) [reply .ruff.toml:14]: It is OK if `ruff format` is present, but seems to be not with `autopep8`.
Let me restore it.

Comment (EarlMilktea) [reply .ruff.toml]: Fixed in 3ac295ef9e11a6f4f315456e0ec6d77db201cb8c .

**Comment by kmaehashi** [comment]
Thanks, LGTM!

**Review by kmaehashi — APPROVED** [review]

**Pull request merged**